File: packages/guards/src/vehicle-id-detect.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface VehicleIdDetectOptions {
  action: 'block' | 'warn';
}

const VIN_PATTERN = /\b[A-HJ-NPR-Z0-9]{17}\b/g;
const US_PLATE = /\b[A-Z]{1,3}\s?\d{1,4}\s?[A-Z]{0,3}\b/;
const EU_PLATE = /\b[A-Z]{1,3}\s?-?\s?\d{1,4}\s?-?\s?[A-Z]{1,3}\b/;

export function vehicleIdDetect(options: VehicleIdDetectOptions): Guard {
  return {
    name: 'vehicle-id-detect',
    version: '0.1.0',
    description: 'Detects VINs and license plate patterns',
    category: 'privacy',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const vins = text.match(VIN_PATTERN) ?? [];
      const hasUsPlate = US_PLATE.test(text);
      const hasEuPlate = EU_PLATE.test(text);
      const vinCount = vins.length;
      const plateCount = (hasUsPlate ? 1 : 0) + (hasEuPlate ? 1 : 0);
      const triggered = vinCount > 0;

      return {
        guardName: 'vehicle-id-detect',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score: triggered ? Math.min((vinCount + plateCount) / 2, 1.0) : 0,
        latencyMs: Math.round(performance.now() - start),
        details: triggered
          ? { vinCount, plateCount }
          : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/vehicle_id_detect.py
"""Detect vehicle identification numbers and plates."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_VIN_PATTERN = re.compile(
    r"\b[A-HJ-NPR-Z0-9]{17}\b"
)
_US_PLATE = re.compile(
    r"\b[A-Z]{1,3}\s?\d{1,4}\s?[A-Z]{0,3}\b"
)
_EU_PLATE = re.compile(
    r"\b[A-Z]{1,3}\s?-?\s?\d{1,4}"
    r"\s?-?\s?[A-Z]{1,3}\b"
)


class _VehicleIdDetect:
    def __init__(
        self, *, action: str = "block"
    ) -> None:
        self.name = "vehicle-id-detect"
        self.action = action

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        vins = _VIN_PATTERN.findall(text)
        vin_count = len(vins)
        triggered = vin_count > 0
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="vehicle-id-detect",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=(
                min(vin_count / 2, 1.0)
                if triggered
                else 0.0
            ),
            message=(
                "Vehicle ID detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"vin_count": vin_count}
                if triggered
                else None
            ),
        )


def vehicle_id_detect(
    *, action: str = "block"
) -> _VehicleIdDetect:
    return _VehicleIdDetect(action=action)
